### PR TITLE
fix: [ANDROAPP-3331] display attributes on carousel event card

### PR DESCRIPTION
**Pull request by mmmateos**
## Description
Please include a summary of the change and include the related jira issue if it exists.

[ jira issue ](https://jira.dhis2.org/browse/ANDROAPP-3331)

## Solution description
If this PR is a fix include a brief description on how the issue is solved.
## Covered unit test cases
Describe the tests that you ran to verify your changes.
## Where did you test this issue?
- [ ] Smartphone Emulator
- [ ] Tablet Emulator
- [ ] Smartphone
- [ ] Tablet
## Which Android versions did you test this issue?
- [ ] 4.4
- [ ] 5.X - 6.X
- [ ] 7.X
- [ ] 8.X
- [ ] 9.X - 10.X
- [ ] Other
## Checklist
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code

### CODE REVIEW

**Review by Balcan — APPROVED** [review]

**Review by ferdyrod — APPROVED** [review]

File: app/src/main/java/org/dhis2/uicomponents/map/carousel/CarouselProgramEventHolder.kt [hunk 1]
```diff
@@ -5,59 +5,128 @@ import android.text.SpannableString
...
+                        } else {
+                            Color.parseColor("#61333333")
+                        }
+                    value.setSpan(
+                        ForegroundColorSpan(colorToUse),
+                        0,
+                        value.length,
+                        Spanned.SPAN_INCLUSIVE_INCLUSIVE
+                    )
+                    stringBuilder.append(value)
```
Comment (JaimeToca) [inline app/src/main/java/org/dhis2/uicomponents/map/carousel/CarouselProgramEventHolder.kt:120]: Do you think it makes sense to extract this presentation logic? This is a very small example but when it gets bigger we cannot unit test :(

Comment (mmmateos) [reply app/src/main/java/org/dhis2/uicomponents/map/carousel/CarouselProgramEventHolder.kt:120]: Okey

File: app/src/main/java/org/dhis2/uicomponents/map/carousel/CarouselProgramEventHolder.kt [hunk 2]
```diff
@@ -1,63 +1,104 @@
...
-            binding.noCoordinatesLabel.noCoordinatesMessage.text =
-                itemView.context.getString(R.string.no_coordinates_item).format(
-                    itemView.context.getString(R.string.event_event)
-                        .toLowerCase(Locale.getDefault())
-                )
+    private fun initValues(
+        valueListIsOpen: Boolean,
+        dataElementValues: MutableList<Pair<String, String>>
+    ) {
+        binding.dataElementList.removeAllViews()
```
Comment (JaimeToca) [inline app/src/main/java/org/dhis2/uicomponents/map/carousel/CarouselProgramEventHolder.kt:72]: why remove and not hide?

File: app/src/main/java/org/dhis2/Bindings/DataElementValuesExtensions.kt [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+            val colorToUse = when {
+                this.indexOf(nameValuePair) % 2 == 0 -> Color.parseColor("#8A333333")
+                else -> Color.parseColor("#61333333")
+            }
+            value.setSpan(
+                ForegroundColorSpan(colorToUse),
+                0,
+                value.length,
+                Spanned.SPAN_INCLUSIVE_INCLUSIVE
+            )
```
Comment (JaimeToca) [inline app/src/main/java/org/dhis2/Bindings/DataElementValuesExtensions.kt:23]: why an extension? I think the presentation logic could be decouple from android dependencies so you can unit test.
Let's talk about it in the daiy :)

File: app/src/main/java/org/dhis2/usescases/programEventDetail/ProgramEventViewModel.java [hunk 4]
```diff
@@ -50,11 +58,33 @@
...
+    }
+
     @NonNull
     public static ProgramEventViewModel create(@NonNull String uid, @NonNull String orgUnitUid, @NonNull String orgUnitName, @NonNull Date date,
                                                @NonNull State eventState, @NonNull List<Pair<String, String>> data, @NonNull EventStatus status,
                                                @NonNull Boolean isExpired, @NonNull String attributeOptionComboName, Geometry geometry, Boolean canBeEdited) {
         return new AutoValue_ProgramEventViewModel(uid, orgUnitUid, orgUnitName, date, eventState, data, status, isExpired, attributeOptionComboName, geometry, canBeEdited);
     }
 
+    public SpannableStringBuilder getSpannableStringValues() {
```
Comment (JaimeToca) [inline app/src/main/java/org/dhis2/usescases/programEventDetail/ProgramEventViewModel.java:74]: can you add a more descriptive name? thanks ! getSpannableString it seems generic and you're getting specific strings

File: app/src/main/java/org/dhis2/usescases/programEventDetail/ProgramEventViewModel.java [hunk 5]
```diff
@@ -50,11 +58,33 @@
...
+                SpannableString value = new SpannableString(nameValuePair.component2());
+                int colorToUse = eventDisplayData().indexOf(nameValuePair) % 2 == 0 ? Color.parseColor("#8A333333") : Color.parseColor("#61333333");
+                value.setSpan(new ForegroundColorSpan(colorToUse), 0, value.length(), Spanned.SPAN_INCLUSIVE_INCLUSIVE);
+                stringBuilder.append(value);
+                if (eventDisplayData().indexOf(nameValuePair) != eventDisplayData().size() - 1) {
+                    stringBuilder.append(" ");
+                }
+            }
+        }
+        return stringBuilder;
```
Comment (JaimeToca) [inline app/src/main/java/org/dhis2/usescases/programEventDetail/ProgramEventViewModel.java:87]: I miss the unit test from both view models :/ . Also, as @andresmr presentation logic should not be couple to android dependencies (you have color here)

File: app/src/main/java/org/dhis2/usescases/teiDashboard/dashboardfragments/teidata/teievents/EventViewModel.kt [hunk 6]
```diff
@@ -23,4 +28,30 @@ data class EventViewModel(
     fun toggleValueList() {
         this.valueListIsOpen = !valueListIsOpen
     }
+
+    fun valuesSpannableString(): SpannableStringBuilder {
+        val stringBuilder = SpannableStringBuilder()
+        dataElementValues?.forEach { nameValuePair ->
+            if (nameValuePair.second != "-") {
+                val value = SpannableString(nameValuePair.second)
```
Comment (JaimeToca) [inline app/src/main/java/org/dhis2/usescases/teiDashboard/dashboardfragments/teidata/teievents/EventViewModel.kt:36]: same here I miss unit test :(

**Comment by JaimeToca** [comment]
Hello @mmmateos very nice job decoupling the logic :clap: :clap: . I added a couple of comments one regarding unit tests, the other was about naming. Let me know what you think.

Thanks a lot

File: app/src/main/java/org/dhis2/usescases/programEventDetail/ProgramEventViewModel.java (hunk 5 above)
Comment (andresmr) [reply app/src/main/java/org/dhis2/usescases/programEventDetail/ProgramEventViewModel.java:87]: Yes, In my opinion, as I comented to @Balcan in a PR related to Strings, I would inject in the **Presenter** or **ViewModel** in this case, a **resourceProvider** with methods like `providePrimaryColor()` and so. Or more granular, different providers (contentProvider, colorProvider ...) as we want to do it.

**Review by JaimeToca — APPROVED** [review]

**Pull request merged**